pkg/server/util/regentryutil/cache.go

```diff
@@ -0,0 +1,66 @@
+package regentryutil
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/spiffe/spire/proto/spire/common"
+)
+
+type RegistrationEntriesCache interface {
+	Get(key string) ([]*common.RegistrationEntry, bool)
+	AddWithExpire(key string, value []*common.RegistrationEntry, expire time.Duration)
+}
+
+// FetchRegistrationEntriesCache is a wrapper around LRU cache with expiry, used for caching registration entries of a agent
+type FetchRegistrationEntriesCache struct {
+	Cache   *lru.Cache
+	TimeNow func() time.Time
+
+	mu sync.RWMutex
+}
+
+type cacheValue struct {
+	entries   []*common.RegistrationEntry
+	expiresAt time.Time
+}
+
+func NewFetchX509SVIDCache(cacheSize int) (*FetchRegistrationEntriesCache, error) {
+	cache, err := lru.New(cacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create lru cache: %v", err)
+	}
+	return &FetchRegistrationEntriesCache{
+		Cache:   cache,
+		TimeNow: time.Now,
+	}, nil
+}
+
+func (c *FetchRegistrationEntriesCache) Get(key string) ([]*common.RegistrationEntry, bool) {
+	c.mu.Lock()
```

Comment:
Considering how much contention there is likely to be on this cache under high load, I think it's important that we use the "read" lock here so we can get many readers at once.

Using the read lock implies that you'll have to do the lookup and check for expiration under the read lock, and if expired, take the write lock and redo the lookup and expiration check to know if you should still expire the entry (since the entry could have been changed while the read lock was released).

Reply:
makes sense. Updated